### PR TITLE
Directives and pipes

### PR DESCRIPTION
**Pull request by gegorov**
In this PR I did the following:

- [ ] created a custom directive to change the border color of course component
- [ ] added no data message using ngIf
- [ ] added "topRated" field and highlight course if it is topRated
- [ ] used a built-in pipe to the uppercase course name
- [ ] created custom duration pipe
- [ ] used built-in date pipe for the course date
- [ ] created custom orderBy pipe and used it to sort courses
- [ ] changed find behavior to filter out user search

### CODE REVIEW

File: src/app/courses-list/course/course.component.spec.ts [hunk 1]
```diff
@@ -1,14 +1,15 @@
 import { async, ComponentFixture, TestBed } from "@angular/core/testing";
 
 import { CourseComponent } from "./course.component";
+import { DurationPipe } from "../../shared/pipes/index";
```
Comment (kirillpanov) [inline src/app/courses-list/course/course.component.spec.ts]: shared/index

File: src/app/courses-list/courses-list.component.html [hunk 2]
```diff
@@ -1,7 +1,21 @@
 <div class="courses-list">
-    <app-search></app-search>
-        <app-course *ngFor="let course of courses$ | async" [course]="course" (removeItem)="onNotify($event)"></app-course>
-    <div class="load-more">
-        <button (click)="loadMore()">Load More</button>
-    </div>
+    <app-search (notify)="onSearchNotify($event)"></app-search>
+<!--    @TODO: get rid of two subscriptions to courses$-->
```
Comment (kirillpanov) [inline src/app/courses-list/courses-list.component.html]: it is ok to have several async pipe subscriptions in the template. So what u have to worry about - your stream should be shared and return last emit to each subscriber.
U can use publishReplay() and refCount(1) rxjs operators

File: src/app/shared/pipes/duration.pipe.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+@Pipe({
+    name: "duration"
+})
+export class DurationPipe implements PipeTransform {
+
+    transform(min: number): string {
+        const MINUTES_IN_HOUR: number = 60;
+
+        if (min < 60) {
```
Comment (kirillpanov) [inline src/app/shared/pipes/duration.pipe.ts]: you have declared constant 2 lines above)

File: src/app/courses-list/courses-list.component.html [hunk 4]
```diff
@@ -1,7 +1,21 @@
...
+<!--    @TODO: and why ngif didn't worked with ng-template -->
+    <ng-container *ngIf="(courses$ | async ).length; else noCoursesFound">
+        <app-course
+            *ngFor="let course of courses$ | async |  filter: filterString | orderBy"
```
Comment (kirillpanov) [inline src/app/courses-list/courses-list.component.html]: so, to use the pipe in the component, it was needed to inject pipe in a component, add the pipe to component's module in the providers sections and use transform method of the pipe inside component's code

File: src/app/shared/pipes/order-by.pipe.spec.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { OrderByPipe } from "./order-by.pipe";
+
+import { COURSES } from "../../core/helpers/mock-courses";
```
Comment (kirillpanov) [inline src/app/shared/pipes/order-by.pipe.spec.ts]: core/index

File: src/app/courses-list/courses-list.component.html [hunk 6]
```diff
@@ -1,7 +1,21 @@
 <div class="courses-list">
-    <app-search></app-search>
-        <app-course *ngFor="let course of courses$ | async" [course]="course" (removeItem)="onNotify($event)"></app-course>
-    <div class="load-more">
-        <button (click)="loadMore()">Load More</button>
-    </div>
+    <app-search (notify)="onSearchNotify($event)"></app-search>
+<!--    @TODO: get rid of two subscriptions to courses$-->
+<!--    @TODO: and why ngif didn't worked with ng-template -->
```
Comment (kirillpanov) [inline src/app/courses-list/courses-list.component.html]: https://stackoverflow.com/questions/44837756/why-ngif-doesntwork-with-ng-template

File: src/app/shared/pipes/order-by.pipe.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import { Pipe, PipeTransform } from "@angular/core";
+import { ICourse } from "../../core/index";
+
+@Pipe({
+    name: "orderBy"
+})
+export class OrderByPipe implements PipeTransform {
+    public transform(courses: Array<ICourse>): Array<ICourse> {
+        return courses.sort((courseA: ICourse, courseB: ICourse) => courseB.creationDate.getTime() - courseA.creationDate.getTime());
```
Comment (kirillpanov) [inline src/app/shared/pipes/order-by.pipe.ts]: it is not a good idea to use sort directly. 
Sort does not return a new array it sorts the source array and mutates it.

File: src/app/shared/pipes/duration.pipe.ts (hunk 3 above)
Comment (gegorov) [reply src/app/shared/pipes/duration.pipe.ts]: fixed :)

File: src/app/courses-list/course/course.component.spec.ts (hunk 1 above)
Comment (gegorov) [reply src/app/courses-list/course/course.component.spec.ts]: fixed

**Pull request merged**